File: src/keywords/process-keywords.mjs
#!/usr/bin/env node
// Deterministic keyword processor: merges raw DataForSEO responses into a
// structured JSON skeleton with intent tags, n-gram clusters, and null
// placeholders for LLM-only fields. Same input always produces byte-identical
// output.
//
// Keyword difficulty is extracted directly from the related_keywords and
// keyword_suggestions responses (keyword_data.keyword_properties.keyword_difficulty),
// eliminating the need for a separate keyword_difficulty/live API call.
//
// Usage:
//   node process-keywords.mjs \
//     --related keywords-related-raw.json \
//     --suggestions keywords-suggestions-raw.json \
//     --seed "keyword recherche" \
//     [--volume keywords-volume-raw.json] \
//     [--brands "brand1,brand2"]

import { readFileSync } from 'node:fs';
import { extractKeywords } from './extract-keywords.mjs';

// --- Parse arguments --------------------------------------------------------

const args = process.argv.slice(2);

function flagValue(name) {
  const idx = args.indexOf(name);
  return idx !== -1 ? args[idx + 1] : undefined;
}

const relatedFile = flagValue('--related');
const suggestionsFile = flagValue('--suggestions');
const volumeFile = flagValue('--volume');
const seedKeyword = flagValue('--seed');
const brandsRaw = flagValue('--brands');

if (!relatedFile || !suggestionsFile || !seedKeyword) {
  console.error(
    'Usage: node process-keywords.mjs --related <file> --suggestions <file> --seed <keyword> ' +
    '[--volume <file>] [--brands "brand1,brand2"]',
  );
  process.exit(1);
}

// --- Intent patterns --------------------------------------------------------
// Word-boundary regex patterns for DE + EN intent classification.
// Order matters: first match wins (transactional > commercial > informational).

const INTENT_PATTERNS = {
  transactional: /\b(kaufen|buy|price|preis|bestellen|coupon|gutschein|discount|rabatt)\b/i,
  commercial: /\b(best|beste|bester|bestes|top|review|vergleich|vs|test|erfahrung|empfehlung)\b/i,
  informational: /\b(how|wie|what|was ist|guide|anleitung|tutorial|tipps|lernen)\b/i,
};

const brandList = brandsRaw
  ? brandsRaw.split(',').map(b => b.trim().toLowerCase()).filter(Boolean)
  : [];

// --- Helpers ----------------------------------------------------------------

function readJSON(path) {
  return JSON.parse(readFileSync(path, 'utf-8'));
}

// extractKeywords is imported from ./extract-keywords.mjs
// It handles both related_keywords and keyword_suggestions response shapes
// and extracts keyword_difficulty from keyword_properties.

// Build a case-insensitive lookup from a separate volume response (optional).
// Expected shape: tasks[0].result[] with keyword, search_volume, cpc.
function buildVolumeMap(raw) {
  const map = new Map();
  const results = raw?.tasks?.[0]?.result;
  if (!Array.isArray(results)) return map;
  for (const item of results) {
    if (item?.keyword != null) {
      map.set(item.keyword.toLowerCase().trim(), {
        search_volume: item.search_volume ?? null,
        cpc: item.cpc ?? null,
      });
    }
  }
  return map;
}

// Classify search intent deterministically.
function classifyIntent(keyword) {
  const lower = keyword.toLowerCase();

  // Check navigational first (brand list)
  if (brandList.length > 0) {
    for (const brand of brandList) {
      if (lower.includes(brand)) return 'navigational';
    }
  }

  // Check patterns in priority order
  if (INTENT_PATTERNS.transactional.test(lower)) return 'transactional';
  if (INTENT_PATTERNS.commercial.test(lower)) return 'commercial';
  if (INTENT_PATTERNS.informational.test(lower)) return 'informational';

  return null;
}

// Tokenize a keyword into lowercase word n-grams (unigrams).
function tokenize(keyword) {
  return keyword.toLowerCase().split(/\s+/).filter(Boolean);
}

// Jaccard similarity between two token sets.
function jaccard(setA, setB) {
  let intersection = 0;
  for (const token of setA) {
    if (setB.has(token)) intersection++;
  }
  const union = setA.size + setB.size - intersection;
  if (union === 0) return 0;
  return intersection / union;
}

// --- Pipeline ---------------------------------------------------------------

// 1. Read all raw JSON files
const relatedRaw = readJSON(relatedFile);
const suggestionsRaw = readJSON(suggestionsFile);
const volumeMap = volumeFile ? buildVolumeMap(readJSON(volumeFile)) : new Map();

// 2. Extract and deduplicate (case-insensitive, trimmed)
const relatedKeywords = extractKeywords(relatedRaw, { includeDifficulty: true });
const suggestionsKeywords = extractKeywords(suggestionsRaw, { includeDifficulty: true });

const seen = new Map(); // lowercase key -> record

for (const kw of relatedKeywords) {
  const key = kw.keyword.toLowerCase().trim();
  if (!seen.has(key)) seen.set(key, kw);
}
for (const kw of suggestionsKeywords) {
  const key = kw.keyword.toLowerCase().trim();
  if (!seen.has(key)) seen.set(key, kw);
}

// Ensure seed keyword is always present
const seedKey = seedKeyword.toLowerCase().trim();
if (!seen.has(seedKey)) {
  seen.set(seedKey, {
    keyword: seedKeyword.trim(),
    search_volume: null,
    cpc: null,
    monthly_searches: null,
  });
}

// 3. Merge volume (from separate endpoint, if provided); difficulty comes
//    from the extracted keyword records (keyword_properties.keyword_difficulty)
const merged = [...seen.values()].map(kw => {
  const key = kw.keyword.toLowerCase().trim();

  // Override volume/CPC from separate volume endpoint if available
  const vol = volumeMap.get(key);
  const search_volume = vol?.search_volume ?? kw.search_volume;
  const cpc = vol?.cpc ?? kw.cpc;

  return {
    keyword: kw.keyword,
    search_volume,
    cpc,
    monthly_searches: kw.monthly_searches,
    difficulty: kw.difficulty ?? null,
  };
});

// Stable sort: volume desc, then alphabetical tie-break
merged.sort((a, b) => {
  const volA = a.search_volume ?? -1;
  const volB = b.search_volume ?? -1;
  if (volB !== volA) return volB - volA;
  return a.keyword.toLowerCase().localeCompare(b.keyword.toLowerCase());
});

// 4. Tag intent
for (const kw of merged) {
  kw.intent = classifyIntent(kw.keyword);
}

// 5. Cluster via n-gram overlap (Jaccard >= 0.5, greedy to highest-volume keyword)
// Keywords are already sorted by volume desc, so first keyword in each cluster
// is the highest-volume one (the cluster representative).

const tokenSets = merged.map(kw => new Set(tokenize(kw.keyword)));
const clusterAssignment = new Array(merged.length).fill(-1);
let nextClusterId = 0;

for (let i = 0; i < merged.length; i++) {
  if (clusterAssignment[i] !== -1) continue;

  // Start a new cluster with this keyword as representative
  const clusterId = nextClusterId++;
  clusterAssignment[i] = clusterId;

  // Greedily assign unassigned keywords with Jaccard >= 0.5
  for (let j = i + 1; j < merged.length; j++) {
    if (clusterAssignment[j] !== -1) continue;
    if (jaccard(tokenSets[i], tokenSets[j]) >= 0.5) {
      clusterAssignment[j] = clusterId;
    }
  }
}

// Group into cluster objects
const clusterMap = new Map(); // clusterId -> { representative index, member indices }
for (let i = 0; i < merged.length; i++) {
  const cid = clusterAssignment[i];
  if (!clusterMap.has(cid)) {
    clusterMap.set(cid, { repIdx: i, members: [] });
  }
  clusterMap.get(cid).members.push(i);
}

// Build cluster array sorted by representative volume desc (already in order
// since we iterate merged in volume-desc order)
const clusters = [];
for (const [, { repIdx, members }] of clusterMap) {
  const representative = merged[repIdx];
  const keywords = members.map(idx => merged[idx]);

  clusters.push({
    cluster_keyword: representative.keyword,
    cluster_label: null, // LLM field
    strategic_notes: null, // LLM field
    keyword_count: keywords.length,
    keywords,
  });
}

// 6. Opportunity score per keyword + re-sort within clusters
// Formula: opportunity_score = search_volume / (keyword_difficulty + 1)
//   - volume null or 0 → score 0
//   - difficulty null → score null
//   - Round to 2 decimal places

function computeOpportunityScore(volume, difficulty) {
  if (difficulty == null) return null;
  if (volume == null || volume === 0) return 0;
  return Math.round((volume / (difficulty + 1)) * 100) / 100;
}

for (const cluster of clusters) {
  // Compute scores
  for (const kw of cluster.keywords) {
    kw.opportunity_score = computeOpportunityScore(kw.search_volume, kw.difficulty);
  }

  // Re-sort: score desc, then volume desc, then alphabetical tie-break
  cluster.keywords.sort((a, b) => {
    const scoreA = a.opportunity_score ?? -1;
    const scoreB = b.opportunity_score ?? -1;
    if (scoreB !== scoreA) return scoreB - scoreA;
    const volA = a.search_volume ?? -1;
    const volB = b.search_volume ?? -1;
    if (volB !== volA) return volB - volA;
    return a.keyword.toLowerCase().localeCompare(b.keyword.toLowerCase());
  });

  // Cluster-level aggregate: average of all scores (nulls count as 0 in sum, but
  // still count toward cluster_size for a fair average)
  const scoreSum = cluster.keywords.reduce(
    (sum, kw) => sum + (kw.opportunity_score ?? 0), 0,
  );
  cluster.cluster_opportunity = cluster.keywords.length > 0
    ? Math.round((scoreSum / cluster.keywords.length) * 100) / 100
    : 0;
}

// 7. Output JSON skeleton
const output = {
  seed_keyword: seedKeyword.trim(),
  total_keywords: merged.length,
  total_clusters: clusters.length,
  clusters,
};

console.log(JSON.stringify(output, null, 2));

File: seo_pipeline/keywords/process_keywords.py
"""Process keywords: intent classification, Jaccard clustering, opportunity scoring.

Deterministic keyword processor that merges raw DataForSEO responses into a
structured JSON skeleton with intent tags, Jaccard clusters, and null
placeholders for LLM-only fields. Same input always produces byte-identical
output.
"""

import argparse
import json
import re
import sys
from pathlib import Path

from seo_pipeline.keywords.extract_keywords import extract_keywords
from seo_pipeline.utils.math import js_round, normalize_number

# Word-boundary regex patterns for DE + EN intent classification.
# Order matters: first match wins (transactional > commercial > informational).
INTENT_PATTERNS = {
    "transactional": re.compile(
        r"\b(kaufen|buy|price|preis|bestellen|coupon|gutschein|discount|rabatt)\b",
        re.IGNORECASE,
    ),
    "commercial": re.compile(
        r"\b(best|beste|bester|bestes|top|review|vergleich|vs|test|erfahrung|empfehlung)\b",
        re.IGNORECASE,
    ),
    "informational": re.compile(
        r"\b(how|wie|what|was ist|guide|anleitung|tutorial|tipps|lernen)\b",
        re.IGNORECASE,
    ),
}


def build_volume_map(raw: dict) -> dict:
    """Build a case-insensitive lookup from a separate volume API response.

    Expected shape: tasks[0].result[] with keyword, search_volume, cpc.

    Args:
        raw: Raw JSON response from the volume endpoint.

    Returns:
        Dict mapping lowercase keyword to {search_volume, cpc}.
    """
    volume_map: dict[str, dict] = {}
    tasks = raw.get("tasks") if raw else None
    if not isinstance(tasks, list) or not tasks:
        return volume_map

    results = tasks[0].get("result") if isinstance(tasks[0], dict) else None
    if not isinstance(results, list):
        return volume_map

    for item in results:
        if not isinstance(item, dict):
            continue
        kw = item.get("keyword")
        if kw is not None:
            volume_map[str(kw).lower().strip()] = {
                "search_volume": item.get("search_volume"),
                "cpc": item.get("cpc"),
            }

    return volume_map


def classify_intent(keyword: str, brand_list: list[str]) -> str | None:
    """Classify search intent deterministically.

    Checks navigational (brand list) first, then pattern-based classification
    in priority order: transactional > commercial > informational.

    Args:
        keyword: The keyword to classify.
        brand_list: List of lowercase brand strings for navigational detection.

    Returns:
        Intent string or None if no match.
    """
    lower = keyword.lower()

    # Check navigational first (brand list)
    if brand_list:
        for brand in brand_list:
            if brand in lower:
                return "navigational"

    # Check patterns in priority order
    if INTENT_PATTERNS["transactional"].search(lower):
        return "transactional"
    if INTENT_PATTERNS["commercial"].search(lower):
        return "commercial"
    if INTENT_PATTERNS["informational"].search(lower):
        return "informational"

    return None


def tokenize_keyword(keyword: str) -> list[str]:
    """Tokenize a keyword into lowercase word tokens (split on whitespace).

    This is the simple tokenizer used for Jaccard similarity, NOT the
    shared tokenizer from utils (which strips punctuation and single chars).

    Args:
        keyword: The keyword to tokenize.

    Returns:
        List of lowercase tokens.
    """
    return [t for t in keyword.lower().split() if t]


def jaccard_similarity(set_a: set[str], set_b: set[str]) -> float:
    """Compute Jaccard similarity between two token sets.

    Args:
        set_a: First set of tokens.
        set_b: Second set of tokens.

    Returns:
        Jaccard coefficient (0.0 to 1.0).
    """
    intersection = len(set_a & set_b)
    union = len(set_a) + len(set_b) - intersection
    if union == 0:
        return 0.0
    return intersection / union


def compute_opportunity_score(
    volume: int | None, difficulty: int | None
) -> float | None:
    """Compute opportunity score for a keyword.

    Formula: search_volume / (keyword_difficulty + 1), rounded to 2 decimals
    using JavaScript Math.round() semantics.

    Args:
        volume: Search volume (None or 0 yields 0).
        difficulty: Keyword difficulty (None yields None).

    Returns:
        Opportunity score, 0, or None.
    """
    if difficulty is None:
        return None
    if volume is None or volume == 0:
        return 0
    return js_round((volume / (difficulty + 1)) * 100) / 100


def process_keywords(
    related_raw: dict,
    suggestions_raw: dict,
    seed: str,
    volume_raw: dict | None = None,
    brands: list[str] | None = None,
) -> dict:
    """Process keywords with intent classification, Jaccard clustering, and scoring.

    Pipeline:
    1. Extract and deduplicate keywords from related + suggestions (with difficulty)
    2. Merge volume from optional volume endpoint
    3. Stable sort by volume desc, alpha tie-break
    4. Classify search intent
    5. Jaccard clustering (greedy, threshold >= 0.5)
    6. Opportunity scoring and within-cluster re-sort
    7. Build output skeleton

    Args:
        related_raw: Raw JSON from related_keywords API.
        suggestions_raw: Raw JSON from keyword_suggestions API.
        seed: Seed keyword string.
        volume_raw: Optional raw JSON from volume endpoint.
        brands: Optional list of brand strings for navigational intent.

    Returns:
        Dict with seed_keyword, total_keywords, total_clusters, clusters.
    """
    brand_list = [b.lower().strip() for b in (brands or []) if b.strip()]

    # Build volume lookup
    volume_map = build_volume_map(volume_raw) if volume_raw else {}

    # 1. Extract and deduplicate (case-insensitive, trimmed)
    related_keywords = extract_keywords(related_raw, include_difficulty=True)
    suggestions_keywords = extract_keywords(suggestions_raw, include_difficulty=True)

    seen: dict[str, dict] = {}  # lowercase key -> record

    for kw in related_keywords:
        key = kw["keyword"].lower().strip()
        if key not in seen:
            seen[key] = kw

    for kw in suggestions_keywords:
        key = kw["keyword"].lower().strip()
        if key not in seen:
            seen[key] = kw

    # Ensure seed keyword is always present
    seed_key = seed.lower().strip()
    if seed_key not in seen:
        seen[seed_key] = {
            "keyword": seed.strip(),
            "search_volume": None,
            "cpc": None,
            "monthly_searches": None,
        }

    # 2. Merge volume from separate endpoint if available; difficulty comes
    # from extracted keyword records
    merged = []
    for kw in seen.values():
        key = kw["keyword"].lower().strip()
        vol = volume_map.get(key)

        search_volume = (
            vol["search_volume"]
            if vol and vol.get("search_volume") is not None
            else kw.get("search_volume")
        )
        cpc = (
            vol["cpc"]
            if vol and vol.get("cpc") is not None
            else kw.get("cpc")
        )

        merged.append({
            "keyword": kw["keyword"],
            "search_volume": search_volume,
            "cpc": normalize_number(cpc),
            "monthly_searches": kw.get("monthly_searches"),
            "difficulty": kw.get("difficulty"),
        })

    # 3. Stable sort: volume desc, then alphabetical tie-break
    merged.sort(
        key=lambda kw: (
            -(kw["search_volume"] if kw["search_volume"] is not None else -1),
            kw["keyword"].lower(),
        )
    )

    # 4. Tag intent
    for kw in merged:
        kw["intent"] = classify_intent(kw["keyword"], brand_list)

    # 5. Cluster via Jaccard overlap (>= 0.5, greedy to highest-volume keyword)
    # Keywords are already sorted by volume desc, so first keyword in each
    # cluster is the highest-volume one (the cluster representative).
    token_sets = [set(tokenize_keyword(kw["keyword"])) for kw in merged]
    cluster_assignment = [-1] * len(merged)
    next_cluster_id = 0

    for i in range(len(merged)):
        if cluster_assignment[i] != -1:
            continue

        # Start a new cluster with this keyword as representative
        cluster_id = next_cluster_id
        next_cluster_id += 1
        cluster_assignment[i] = cluster_id

        # Greedily assign unassigned keywords with Jaccard >= 0.5
        for j in range(i + 1, len(merged)):
            if cluster_assignment[j] != -1:
                continue
            if jaccard_similarity(token_sets[i], token_sets[j]) >= 0.5:
                cluster_assignment[j] = cluster_id

    # Group into cluster objects
    cluster_map: dict[int, dict] = {}  # cluster_id -> {rep_idx, members}
    for i in range(len(merged)):
        cid = cluster_assignment[i]
        if cid not in cluster_map:
            cluster_map[cid] = {"rep_idx": i, "members": []}
        cluster_map[cid]["members"].append(i)

    # Build cluster array (already in order since we iterate merged in volume-desc)
    clusters = []
    for info in cluster_map.values():
        representative = merged[info["rep_idx"]]
        keywords = [merged[idx] for idx in info["members"]]

        clusters.append({
            "cluster_keyword": representative["keyword"],
            "cluster_label": None,
            "strategic_notes": None,
            "keyword_count": len(keywords),
            "keywords": keywords,
        })

    # 6. Opportunity score per keyword + re-sort within clusters
    for cluster in clusters:
        # Compute scores
        for kw in cluster["keywords"]:
            kw["opportunity_score"] = compute_opportunity_score(
                kw["search_volume"], kw["difficulty"]
            )

        # Re-sort: score desc, then volume desc, then alphabetical tie-break
        cluster["keywords"].sort(
            key=lambda kw: (
                -(
                    kw["opportunity_score"]
                    if kw["opportunity_score"] is not None
                    else -1
                ),
                -(kw["search_volume"] if kw["search_volume"] is not None else -1),
                kw["keyword"].lower(),
            )
        )

        # Cluster-level aggregate: average of all scores
        # (nulls count as 0 in sum, but still count toward cluster size)
        score_sum = sum(
            kw["opportunity_score"] if kw["opportunity_score"] is not None else 0
            for kw in cluster["keywords"]
        )
        cluster["cluster_opportunity"] = normalize_number(
            js_round((score_sum / len(cluster["keywords"])) * 100) / 100
            if cluster["keywords"]
            else 0
        )

    # 7. Output JSON skeleton
    return {
        "seed_keyword": seed.strip(),
        "total_keywords": len(merged),
        "total_clusters": len(clusters),
        "clusters": clusters,
    }


def main() -> None:
    """CLI wrapper for process_keywords."""
    parser = argparse.ArgumentParser(
        description="Process keywords with intent, Jaccard clustering, and scoring"
    )
    parser.add_argument(
        "--related", required=True, help="Path to related_keywords raw JSON file"
    )
    parser.add_argument(
        "--suggestions",
        required=True,
        help="Path to keyword_suggestions raw JSON file",
    )
    parser.add_argument(
        "--seed", required=True, help="Seed keyword"
    )
    parser.add_argument(
        "--volume", default=None, help="Path to volume raw JSON file (optional)"
    )
    parser.add_argument(
        "--brands", default=None, help="Comma-separated brand list (optional)"
    )
    parser.add_argument(
        "--output", default=None, help="Output file path (default: stdout)"
    )

    args = parser.parse_args()

    try:
        related_path = Path(args.related)
        suggestions_path = Path(args.suggestions)

        if not related_path.exists():
            print(f"Error: related file not found: {args.related}", file=sys.stderr)
            sys.exit(1)
        if not suggestions_path.exists():
            print(
                f"Error: suggestions file not found: {args.suggestions}",
                file=sys.stderr,
            )
            sys.exit(1)

        with open(related_path, encoding="utf-8") as f:
            related_raw = json.load(f)
        with open(suggestions_path, encoding="utf-8") as f:
            suggestions_raw = json.load(f)

        volume_raw = None
        if args.volume:
            volume_path = Path(args.volume)
            if not volume_path.exists():
                print(
                    f"Error: volume file not found: {args.volume}", file=sys.stderr
                )
                sys.exit(1)
            with open(volume_path, encoding="utf-8") as f:
                volume_raw = json.load(f)

        brands = None
        if args.brands:
            brands = [b.strip() for b in args.brands.split(",") if b.strip()]

        result = process_keywords(
            related_raw, suggestions_raw, args.seed,
            volume_raw=volume_raw, brands=brands,
        )

        output_json = json.dumps(result, indent=2)
        if args.output:
            Path(args.output).write_text(output_json, encoding="utf-8")
        else:
            print(output_json)

    except json.JSONDecodeError as e:
        print(f"Error: invalid JSON: {e}", file=sys.stderr)
        sys.exit(1)


if __name__ == "__main__":
    main()
